src/main/java/net/minecraftforge/fluids/FluidIngredient.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Minecraft Forge
+ * Copyright (c) 2016.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.fluids;
+
+import com.google.common.base.Preconditions;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonSyntaxException;
+import it.unimi.dsi.fastutil.ints.IntList;
+import it.unimi.dsi.fastutil.ints.IntLists;
+import net.minecraft.item.ItemStack;
+import net.minecraft.item.crafting.Ingredient;
+import net.minecraft.nbt.JsonToNBT;
+import net.minecraft.nbt.NBTException;
+import net.minecraft.util.JsonUtils;
+import net.minecraftforge.common.crafting.JsonContext;
+
+import javax.annotation.Nonnull;
+import java.util.Locale;
+
+public class FluidIngredient extends Ingredient
+{
+
+    public enum MatchingStrategy
+    {
+        /**
+         * EXACT matching indicates that it will try matching exact amount of fluid.
+         * Example: if the FluidIngredient asks for 500 mB water, a vanilla water bucket
+         * will not be matched, as it holds 1000 mB water.
+         */
+        EXACT,
+
+        /**
+         * EXCEED matching is the default matching strategy, when not specified.
+         * It will try matching those containers that can drain the specified amount of
+         * fluid out. Example: if the FluidIngredient asks for 500 mB water, a vanilla bucket
+         * will be matched.
+         */
+        EXCEED
+    }
+
+    private final FluidStack fluidStack;
+    private final MatchingStrategy strategy;
+
+    public FluidIngredient(FluidStack fluidStack, MatchingStrategy strategy)
+    {
+        super(0);
+        Preconditions.checkNotNull(fluidStack, "Null FluidStack is not permitted in FluidIngredient");
+        Preconditions.checkNotNull(strategy, "MatchingStrategy cannot be null");
+        this.fluidStack = fluidStack;
+        this.strategy = strategy;
+    }
+
+    public FluidStack getFluidStack()
+    {
+        return fluidStack.copy();
+    }
+
+    public MatchingStrategy getMatchingStrategy()
+    {
+        return strategy;
+    }
+
+    @Override
+    public ItemStack[] getMatchingStacks()
+    {
+        return new ItemStack[0];
```

Comment:
We might need some improvement on this one, I guess...

Reply:
>  I have considered replacing it with ItemStack[1] that contains a forge universal bucket, but not sure whether it is appropriate or not. Inputs are welcome.

Reply:
Actually no, the universal bucket is not enabled by default...

Reply:
@mezz Is there any chance to add a new Item for displaying `FluidStack` here, or I have to patch vanilla recipe book to allow `ItemStack[0]` and `IntLists.EMPTY`?

Reply:
It would be more user friendly to see the items that hold the fluid here.
I am not sure what the most efficient way to find those items is in Forge, but I think it's worth looking into.

Reply:
is it a a option to bring fluidBucketRegistry back ??

Reply:
You mean the `FluidContainerRegistry`? I am afraid that it's impossible.

Reply:
We could listen to `AttachCapabilitiesEvent<ItemStack>` to check every itemstack with capabilities and create a list of everything that has the fluid item capability. Beyond that I don't really have a good idea of how to do it.

Reply:
Does that guarantee that we can get at least 1 valid container?

Reply:
If a fluid recipe is registered with no valid container to work with it, that is an issue. You could hide it from the recipe book in that case (using `IRecipe#isDynamic`?) since it is an uncraftable recipe.

Reply:
Will see if that works. Second question: doesn't maintaining a collection of known containers consume memory?

Reply:
It does, but not a lot.

Reply:
How would that work with items that dynamically expose the fluid capability? Such as an item which can hold liquid if there is a certain upgrade applied? Chances are that the cap won't be exposed during AttachCapabilitiesEvent.
Having an additional way to add itemstacks to the list could be useful (could add an itemstack with the upgrade applied).

Reply:
For simplicity I don't think this would support dynamically exposed fluid capabilities. It's just meant to give the player a general idea of what can be used in the recipe.

Reply:
I have an idea: specify that as a part of `FluidIngredient`? Like this:

````json
{
  "type": "forge:fluid",
  "fluid": "minecraft:water",
  "amount": 1000,
  "matching": "exact",
  "container": "forge:bucket"
}

Reply:
I would prefer the more automatic way, it can show most valid items without any effort.

Reply:
@3TUSK Adding it to the FluidIngredient would mean that some people might think that it had to be a forge:bucket in said recipe, rather than any container

Reply:
I think this is great now.